draft-ietf-dnssd-privreq.txt

```diff
@@ -777,71 +850,190 @@ Internet-Draft         DNS-SD Privacy Requirements            March 2018
 
+   5.  No fingerprinting: Assuming servers use fresh nonces per private
+       instance name, advertisements change regularly.
 
+A.2.  Private IoT
 
+   Boneh et al.  [Wu16] developed an approach for private service
+   discovery that reduces to private mutual authentication.  Moreover,
+   it should be infeasible for any adversary to forge advertisements or
+   impersonate anyone else on the network.  Specifically, service
+   discoverers only wish to reveal their identity to services they
+   trust, and vice versa.  Existing protocols such as TLS, IKE, and
+   SIGMA [SIGMA] require that one side reveal its identity first.  Their
+   approach first allocates, via some policy manager, key pairs
+   associated with human-readable policy names.  For example, user Alice
+   might have a key pair associated with the names /Alice, /Alice/
+   Family, and /Alice/Device.  Her key is bound to each of these names.
+   Authentication policies (and trust models) are then expressed as
+   policy prefix patterns, e.g., /Alice/*. Broadcast messages are
+   encrypted to policies.  For example, Alice might encrypt a message m
+   to the policy /Bob/*. Only Bob, who owns a private key bound to,
+   e.g., /Bob/Devices, can decrypt m.  (This procedure uses a form of
+   identity-based encryption called prefix-based encryption.  Readers
+   are referred to [Wu16] for a thorough description.)
 
+   Using prefix- and policy-based encryption, service discovery is
+   decomposed into two steps: (1) service announcement and (2) key
+   exchange, similar to [I-D.huitema-dnssd-privacy].  Announcements
+   carry service identities, ephemeral key shares, and a signature, all
+   encrypted under the service's desired policy prefix, e.g., /Alice/
+   Family/*. Upon receipt of an announcement, clients with matching
 
 
 
-Huitema, et al.        Expires September 25, 2018              [Page 15]
+Huitema, et al.        Expires September 28, 2018              [Page 18]
 
 Internet-Draft         DNS-SD Privacy Requirements            March 2018
 
 
+   policy private keys can decrypt the announcement and use the
+   ephemeral key share to perform an Authenticated Diffie Hellman key
+   exchange with the service.  Upon completion, the derived shared
+   secret may be used for any further communication, e.g., DNS-SD
+   queries, if needed.
+
+   This system has the following properties:
+
+   1.  Asymmetric work load: computation for clients is on the order of
+       advertisements.
+
+   2.  Mutual identity privacy: Both client and server identities are
+       hidden from active and passive attackers.
+
+   3.  Client set size hiding: Policy-based encryption advertisements
+       hides the number of clients with matching policy keys.
+
+   4.  Unlinkability: Client initiated connections are unlinkable to
+       service advertisements (modulo network-layer connection
+       information, such as advertisement origin and connection
+       destination).
+
```

Comment:
There is another issue there, the real time requirement. The discovery starts with the client encoding a message with the server's public key; fancy crypto ensures that the public key is not revealed when doing that. The server then decodes the message with its private key, verifies that the client is properly authorized, and replies using a key that is private to client and server.

Contrast that with the "DNS" part of DNS-SD, in which the server posts a set of DNS records to a server, and the client later retrieves some of those records. The records are posted before the server knows which client will try retrieve them. We need to have a discussion of whether this "post then retrieve" mode of operation is required, or whether we can ditch it.

Reply:
I filed an issue to discuss the real time requirement.